File: 2024-js/day1.js
let input = document.body.innerText.split('\n').map(line => line.split('   '));
let left = [], right = [], diff = 0, similarity = 0;

input.forEach(n => left.push(n[0]) && right.push(n[1]));

left.pop();
right.pop();

left.sort().map(n => parseInt(n));
right.sort().map(n => parseInt(n));

for (let i in left) {
  let a = left[i], b = right[i];
  diff += Math.abs(a - b);
  similarity += a * right.filter(k => k === a).length;
}

console.log("Diff:", diff);
console.log("Similarity:", similarity);

File: 2024-py/day1.py
from collections import defaultdict

with open("2024-inputs/day1.txt", "r") as f:
    contents = f.read().splitlines()

lines = [x.split("   ") for x in contents]
left = [int(a) for (a, _) in lines]
right = [int(b) for (_, b) in lines]

left.sort()
right.sort()

total = 0
similarity = 0

counts = defaultdict(int)

for a in right:
    counts[a] += 1

for i, a in enumerate(left):
    b = right[i]

    diff = abs(a - b)
    total += diff

    similarity += a * counts[a]

print("Part 1: Total diff:", total)
print("Part 2: similarity:", similarity)
